---
title: "Report"
output: html_document
---

Run MiXCR on raw seq data.

```{bash eval=FALSE, engine.opts='-l', include=FALSE}
mixcr analyze amplicon --species hsa --starting-material dna --5-end no-v-primers --3-end c-primers --adapters no-adapters "data/BSSE_QGF_115935_000000000_CBD4K_1_mHER_H3_3_AgN_AGTCAA_S13_L001_R1_001_MM_1.fastq" "data/BSSE_QGF_115935_000000000_CBD4K_1_mHER_H3_3_AgN_AGTCAA_S13_L001_R2_001_MM_1.fastq" MiXCR/analysis_1

mixcr analyze amplicon --species hsa --starting-material dna --5-end no-v-primers --3-end c-primers --adapters no-adapters "data/BSSE_QGF_115936_000000000_CBD4K_1_mHER_H3_3_2Ag647_AGTTCC_S14_L001_R1_001_MM_1.fastq" "data/BSSE_QGF_115936_000000000_CBD4K_1_mHER_H3_3_2Ag647_AGTTCC_S14_L001_R2_001_MM_1.fastq" MiXCR/analysis_2

mixcr analyze amplicon --species hsa --starting-material dna --5-end no-v-primers --3-end c-primers --adapters no-adapters "data/BSSE_QGF_115937_000000000_CBD4K_1_mHER_H3_3_2Ag488_ATGTCA_S15_L001_R1_001_MM_1.fastq" "data/BSSE_QGF_115937_000000000_CBD4K_1_mHER_H3_3_2Ag488_ATGTCA_S15_L001_R2_001_MM_1.fastq" MiXCR/analysis_3

```


To run IgBLAST fasta sequences for V,D,J germline genes have to be downloaded from IGMT.

```{r}
v_genes_url <- "http://www.imgt.org/download/V-QUEST/IMGT_V-QUEST_reference_directory/Homo_sapiens/IG/IGHV.fasta"
d_genes_url <- "http://www.imgt.org/download/V-QUEST/IMGT_V-QUEST_reference_directory/Homo_sapiens/IG/IGHD.fasta"
j_genes_url <- "http://www.imgt.org/download/V-QUEST/IMGT_V-QUEST_reference_directory/Homo_sapiens/IG/IGHJ.fasta"

download.file(v_genes_url, "igblast/database/IGHV.fasta", quiet = FALSE, mode = "w", headers = NULL)
download.file(d_genes_url, "igblast/database/IGHD.fasta", quiet = FALSE, mode = "w", headers = NULL)
download.file(j_genes_url, "igblast/database/IGHJ.fasta", quiet = FALSE, mode = "w", headers = NULL)
```


```{bash eval=FALSE, engine.opts='-l', include=FALSE}
cd igblast

bin/edit_imgt_file.pl database/IGHV.fasta > IGHV.db
bin/edit_imgt_file.pl database/IGHD.fasta > IGHD.db
bin/edit_imgt_file.pl database/IGHJ.fasta > IGHJ.db

bin/makeblastdb -parse_seqids -dbtype nucl -in IGHV.db
bin/makeblastdb -parse_seqids -dbtype nucl -in IGHD.db
bin/makeblastdb -parse_seqids -dbtype nucl -in IGHJ.db

bin/igblastn -germline_db_V IGHV.db -germline_db_J IGHJ.db -germline_db_D IGHD.db -query myseq/BSSE_QGF_115935_000000000_CBD4K_1_mHER_H3_3_AgN_AGTCAA_S13_L001_R1_001_MM_1.fastq -auxiliary_data -show_translation

bin/igblastn -germline_db_V IGHV.db -germline_db_J IGHJ.db -germline_db_D IGHD.db -query ../data/BSSE_QGF_115935_000000000_CBD4K_1_mHER_H3_3_AgN_AGTCAA_S13_L001_R1_001_MM_1.fastq -auxiliary_data -show_translation -remote
```



```{r reading MiXCR output, message=FALSE}
library(tidyverse, quietly = T)
require(ggplot2); theme_set(theme_bw())

# If labeled by aa or nt CDR3 sequence
aa <- T

# reading output of MiXCR alignment
neg <- read.table("MiXCR/analysis_1.clonotypes.IGH.txt", header = T, sep = "\t")
pos_1 <- read.table("MiXCR/analysis_2.clonotypes.IGH.txt", header = T, sep = "\t")
pos_2 <- read.table("MiXCR/analysis_3.clonotypes.IGH.txt", header = T, sep = "\t")

pos <- bind_rows(pos_1, pos_2)

filtered_positive <- pos %>% filter(case_when(
        aa ~!aaSeqCDR3 %in% neg$aaSeqCDR3,
        !aa ~!nSeqCDR3 %in% neg$nSeqCDR3)) %>% mutate(label = 1)

filtered_negative <- neg %>% filter(case_when(
        aa ~!aaSeqCDR3 %in% pos$aaSeqCDR3,
        !aa ~!nSeqCDR3 %in% pos$nSeqCDR3)) %>% mutate(label = 0)

features <- bind_rows(filtered_negative, filtered_positive) %>%
    select_if(~sum(!is.na(.)) > 0) %>%
        distinct(case_when(aa ~ aaSeqCDR3,
                          !aa ~ nSeqCDR3), .keep_all = T) %>% select(-matches("case_when"))


# Filtering only accepted sequences
cdr3_pattern <- "CSR[FWY][ADEGHIKLMNPQRSTV][ADEGHIKLMNPQRSTV][ACDFGHILNPRSTVY][GS][FLM]Y[AEKLMPQTV][FHILMNY][ADEHIKLMNPQTV]YW"

features <- features %>% filter(str_detect(aaSeqCDR3, cdr3_pattern))

write.csv(features, "data/feat_2.csv", row.names = F)
```

Checking CDR3 length ditribution.

```{r repertoire, echo=FALSE}
library(caret, quietly = T)
features <- read.csv("data/feat_2.csv")
features$label <- as.factor(features$label)

# Clonal expansion 
ggplot(filter(features, cloneId < 1000), aes(x=cloneId, y=cloneFraction, fill=label)) + geom_col()
table(features$label)

# Balancing dataset if needed
# features <- upSample(select(features, -label), features$label, yname = "label")

table(features$label)

write.csv(features, "data/feat_2.csv", row.names = F)

# CDR3 length 
features <- features %>% mutate(cdr3_len = str_length(aaSeqCDR3), mean_len = mean(cdr3_len))
features %>% ggplot(aes(x=cdr3_len, fill=label)) + geom_histogram()
    

```
Filtering CDR3 lengths
```{r}
features <- features %>% filter(cdr3_len < 20)
```


Distribution of levenstein distances to Trastuzumab.

```{r}
library(stringdist)
cdr3_drug <- "CSRWGGDGFYAMDYW"
features$lv_dist <- stringdist::stringdist(features$aaSeqCDR3, cdr3_drug, c("lv"))
ggplot(features, aes(x=lv_dist, fill=label)) + geom_histogram()

```


Selecting only sequences that differ from Trastuzumab CDR3 of at least n positions and not more than m.

```{r}
min_lv_dist <- 3
max_lv_dist <- 10

features <- features %>% filter(lv_dist > min_lv_dist, lv_dist < max_lv_dist)

write.csv(features, "data/feat_2.csv", row.names = F)
```

```{r echo=F}
source("ML_visual.R")
runApp("ML_visual.R")
```

Exploring feature importance (unfinished).

```{r}
if (file.exists("data/feat_imp.csv")) {
    imp = read.csv("data/feat_imp.csv")
}

imp <- imp %>% mutate(pos = parse_number(name), aa = gsub("[^[:upper:]]", "", name)) %>% select(-name) %>% group_split(pos)

aa <- c("G", "P", "A", "V", "L", "I", "M", "C", "F", "Y", "W","H","K","R","Q","N", "E","D","S","T")

b <- data.frame(aa)

try <- map(a, ~complete(select(., c(value, aa)), b)) %>% reduce(left_join, by = "aa") %>% 
    replace(is.na(.), 0) %>% select(-aa)

boxplot(try, use.cols = TRUE)
```
